astropy/cosmology/tests/test_core.py

```diff
@@ -0,0 +1,325 @@
+        assert newparam.fget == "NOT NONE"
+
+    # -------------------------------------------
+    # misc
+
+    def test_repr(self, cosmo_cls):
+        r = repr(cosmo_cls.param)
+        assert f"Parameter 'param'" in r
+        assert str(hex(id(cosmo_cls.param))) in r
+
+    # ==============================================================
+
+    def test_parameter_doesnt_change_with_generic_class(self):
+        """Descriptors are initialized once and not updated on subclasses."""
+
+        class ExampleBase:
+            def __init__(self, param=15):
+                self._param = param
+
+            sig = inspect.signature(__init__)
+            _init_signature = sig.replace(parameters=list(sig.parameters.values())[1:])
+
+            param = Parameter(doc="example parameter")
+
+        class Example(ExampleBase): pass
+
+        assert Example.param is ExampleBase.param
+
+    def test_parameter_doesnt_change_with_cosmology(self, cosmo_cls):
+        """Cosmology reinitializes all descriptors when a subclass is defined."""
+
+        # define subclass to show param is different
+        class Example(cosmo_cls): pass
+
+        assert Example.param is cosmo_cls.param
+
+        # unregister
+        _COSMOLOGY_CLASSES.pop(Example.__qualname__)
+        assert Example.__qualname__ not in _COSMOLOGY_CLASSES
+
+
+# ========================================================================
+
+
+class ParameterTestMixin:
+    """Tests for a Parameter on a Cosmology."""
+
+    def test_Parameters(self, cosmo_cls):
+        """Test `astropy.cosmology.Parameter` attached to Cosmology."""
+        # first establish has expected attribute
+        assert hasattr(cosmo_cls, "__parameters__")
+
+        # check that each entry is a Parameter
+        for n in cosmo_cls.__parameters__:
+            assert hasattr(cosmo_cls, n)  # checks on the instance
+            assert isinstance(getattr(cosmo_cls, n), Parameter)
+
+        # the reverse: check that if it is a Parameter, it's listed.
+        for n in dir(cosmo_cls):
+            if isinstance(getattr(cosmo_cls, n), Parameter):
+                assert n in cosmo_cls.__parameters__
+
+    def test_Parameter_not_unique(self, cosmo_cls, clean_registry):
+        """Cosmology reinitializes Parameter when a class is defined."""
+
+        # define subclass to show param is different
+        class ExampleBase(cosmo_cls):
+            param = Parameter()
+
+        class Example(ExampleBase): pass
+
+        assert Example.param is ExampleBase.param
+        assert Example.__parameters__ == ExampleBase.__parameters__
+
+    def test_Parameters_reorder_by_signature(self, cosmo_cls, clean_registry):
+        """Test parameters are reordered."""
+
+        class Example(cosmo_cls):
+            param = Parameter()
+
+            def __init__(self, param, *, name=None, meta=None):
+                pass  # never actually initialized
+
+        # param should be 1st, all other parameters next
```

Comment:
Check the other parameters are as expected.

Reply:
✅